### PR TITLE
Add nullability marker annotations

### PR DESCRIPTION
**Pull request by TheMrMilchmann**
As previously discussed in Slack, this PR adds JSR-305 like annotations to support null-analysis and to enhance Kotlin interopertability. (See [here](https://kotlinlang.org/docs/reference/java-interop.html#jsr-305-support) for more info.)

@Spasi As suggested I did not add `@Nullable` to most stack allocation methods (the exception being String conversion methods, since they can also be `null` when the passed `CharSequence` is `null`).
I'm open for further suggestions and will implement them ASAP if requested.

### CODE REVIEW

**Comment by TheMrMilchmann** [comment]
I just cloned this branch into a fresh environment and realized that the latest commits seem to be missing some changes. I'll quickly rebase and fix this.

**Comment by TheMrMilchmann** [comment]
Rebased and updated. Should be good to go now (this time for real)!

File: doc/notes/3.1.6.md [hunk 1]
```diff
@@ -21,6 +21,10 @@ This build includes the following changes:
 
 #### Improvements
 
+- Added additional nullability information with [JSR-305](https://jcp.org/en/jsr/detail?id=305) annotations.
+    * Enables null-analysis to be used to improve the code quality of LWJGL applications.
+    * Improves type-safety for applications using LWJGL in Kotlin
+    ([Read the Kotlin documentation](https://kotlinlang.org/docs/reference/java-interop.html#jsr-305-support) for more info.)
```
[the remark this answers is not in the document]
Comment (Spasi) [reply doc/notes/3.1.6.md]: I have changed the wording a bit.

File: modules/core/src/main/java/org/lwjgl/system/MemoryStack.java [hunk 2]
```diff
@@ -531,23 +532,31 @@ public ByteBuffer ASCII(CharSequence text) {
     /**
      * Encodes the specified text on the stack using ASCII encoding and returns a ByteBuffer that points to the encoded text.
      *
-     * @param text           the text to encode. If {@code text} is null, null is returned.
+     * @param text           the text to encode
      * @param nullTerminated if true, a null-terminator is included at the end of the encoded text
      */
     public ByteBuffer ASCII(CharSequence text, boolean nullTerminated) {
```
[the remark this answers is not in the document]
Comment (Spasi) [reply modules/core/src/main/java/org/lwjgl/system/MemoryStack.java]: There's nothing special about the encoding methods, they fail in the same way as other stack allocation methods with an `OutOfMemoryError("Out of stack space.")`.

File: modules/core/src/main/java/org/lwjgl/system/MemoryUtil.java [hunk 3]
```diff
@@ -442,6 +493,19 @@ public static long nmemAlignedAlloc(long alignment, long size) {
         return ALLOCATOR.aligned_alloc(alignment, size);
     }
 
+    public static long nmemAlignedAllocChecked(long alignment, long size) {
```
[the remark this answers is not in the document]
Comment (Spasi) [reply modules/core/src/main/java/org/lwjgl/system/MemoryUtil.java]: Fixed.

File: modules/generator/src/main/kotlin/org/lwjgl/generator/Types.kt [hunk 4]
```diff
@@ -365,6 +367,9 @@ val NativeType.isPointer
 val NativeType.isPointerData
     get() = this is PointerType && this.mapping !== PointerMapping.OPAQUE_POINTER
 
+val NativeType.isPointerObject
+    get() = this is PointerType && (this.mapping !== PointerMapping.OPAQUE_POINTER || this is ObjectType)
```
[the remark this answers is not in the document]
Comment (Spasi) [reply modules/generator/src/main/kotlin/org/lwjgl/generator/Types.kt]: Thanks! I refactored some aspects of the type system a bit and this was simplified to:

```kotlin
this !is ValueType && (this.mapping !== PointerMapping.OPAQUE_POINTER || this is ObjectType)
```

The property was also renamed to `isReference`, which I think is more appropriate.

**Comment by Spasi** [comment]
The next snapshot (`3.1.6 build 9`) will have all the latest updates in bindings, but not the nullability annotations. It will be promoted to stable when it's up. Starting from `build 10`,  the above changes will be available for testing.

Thank you @TheMrMilchmann for the great work!

**Pull request merged**